### PR TITLE
[PT FE] Added support for torch_scatter::segment_mean_csr

### LINKED ISSUE #29734: [Good First Issue]: Support torch_scatter::segment_mean_csr
### Context

Example model to reproduce operation:
```
from torch_scatter import segment_csr
import torch


class Example(torch.nn.Module):
    def forward(self, src, indptr):
        return segment_csr(src, indptr, reduce="mean")

src = torch.randn(10, 6, 64)
indptr = torch.tensor([0, 2, 5, 6])
indptr = indptr.view(1, -1)

m = Example()
t = torch.jit.trace(m, example_inputs=(src, indptr))
print(t.inlined_graph)
```

```
graph(%self : __torch__.Example,
      %src : Float(10, 6, 64, strides=[384, 64, 1], requires_grad=0, device=cpu),
      %indptr : Long(1, 4, strides=[4, 1], requires_grad=0, device=cpu)):
  %3 : NoneType = prim::Constant()
  %4 : Float(10, 3, 64, strides=[192, 64, 1], requires_grad=0, device=cpu) = torch_scatter::segment_mean_csr(%src, %indptr, %3) # /home/vmaxim/openvino/ov_env/lib/python3.10/site-packages/torch/_ops.py:1123:0
  return (%4)
```

### What needs to be done?

As in the main issue.

### Example Pull Requests

_No response_

### Resources

- [torch_scatter.segment_csr](https://pytorch-scatter.readthedocs.io/en/latest/functions/segment_csr.html)
- [Contribution guide - start here!](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [Intel DevHub Discord channel](https://discord.gg/7pVRxUwdWG) - engage in discussions, ask questions and talk to OpenVINO developers
- [How to link your Pull Request to an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue-using-the-pull-request-sidebar)

### Contact points

As in the main issue.

### Ticket

_No response_

### PR DESCRIPTION
**Pull request by Nishant-ZFYII**
**Closes #29734**

### Details

* Implemented PyTorch Frontend translator for `torch_scatter::segment_mean_csr` using a **prefix-sum + gather boundaries** approach (no unrolling over segments).
* The segmentation axis is computed dynamically as **`axis = indptr.dim() - 1`** (matches torch_scatter semantics).
* Added handling for **batch/broadcast-style indptr** (when `axis > 0`) via bidirectional broadcast to `[src_batch_dims..., indptr_last_dim]`.
* Uses **i64 indices** for `indptr`/shape operations to avoid overflow and to match Gather-supported index types.
* Handles the optional `out` argument via `context.mutate_input(out_idx, result)` (if provided).
* Added a regression test for TorchScript that covers:

  * 2D `indptr` with shape `(1, n)` (segments along dim 1)
  * 1D `indptr` (segments along dim 0)

### Limitations / Notes

* Requires **static ranks** for `src` and `indptr` (validated).
* Empty segments (`indptr[i+1] == indptr[i]`) are handled by clamping length to 1 to avoid division by zero (result becomes 0 for empty segments).
* Test is **TorchScript-only** and is skipped for torch.export / torch.compile backends and when `torch_scatter` is not installed.

### How to test

* PyTorch layer test:

  * `python -m pytest tests/layer_tests/pytorch_tests/test_segment_mean_csr.py -k segment_mean_csr -v`

### Tickets

* #29734
* Parent tracking: #28584

### CODE REVIEW

**Comment by Nishant-ZFYII** [comment]
@mvafin Can you please review . 

Thanks. !

**Comment by Nishant-ZFYII** [comment]
@mlukasze @p-wysocki 

Good Day, 

Kindly requesting you to review the PR. Please let me know if further changes are required. 

Thanks and Regards.

**Comment by Nishant-ZFYII** [comment]
Good Day @mmikolajcz ,  @PiotrKrzem :

Kindly requesting you to review the PR. Please let me know if further changes are required.

Thanks and Regards, 
Nishant.

**Comment by Nishant-ZFYII** [comment]
Good Day @mmikolajcz  , @PiotrKrzem  :

Kindly requesting you to review the PR. Please let me know if further changes are required.

Thanks and Regards,
Nishant.

**Comment by Nishant-ZFYII** [comment]
Good Day @mmikolajcz  , @PiotrKrzem  :

Kindly requesting you to review the PR. Please let me know if further changes are required.

Thanks and Regards,
Nishant.

**Comment by Nishant-ZFYII** [comment]
Good Day @mmikolajcz , @PiotrKrzem :

Kindly requesting you to review the PR. Please let me know if further changes are required.

Thanks and Regards,
Nishant.

**Comment by Nishant-ZFYII** [comment]
@openvinotoolkit/openvino-pytorch-frontend-maintainers, @mvafin 

Kindly requesting you to review the PR. Please let me know if further changes are required.

Thanks and Regards,
Nishant.

**Comment by Nishant-ZFYII** [comment]
@openvinotoolkit/openvino-pytorch-frontend-maintainers, @mvafin

Kindly requesting you to review the PR. Please let me know if further changes are required.

Thanks and Regards,
Nishant.

**Comment by Nishant-ZFYII** [comment]
I've made the requested variable name changes. 
I also noticed that the tests were failing when I ran in my local machine. 
I made the below changes: 

1. Implemented segmented mean using a prefix-sum flow (CumSum, Gather, Subtract, Divide).
2. Added safe handling for optional out input.
3. Converted indptr to i64 for robust index handling.
4. Added dynamic-rank support for src.
5. Fixed the broadcast/rank issue by using ShapeOf(ShapeOf(src)) when computing trailing dims.
6. Kept empty segment handling stable by clamping lengths before division.

It passed 8 cpu tests and need to pass GPU tests .The  GPU failures are environment-related (GPU device not registered), not implementation-related.

Please let me know if any more changes are required. 

Happy to help :)

File: src/frontends/pytorch/src/op/segment_mean_csr.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+            context.mark_node(std::make_shared<v0::Concat>(OutputVector{batch_shape, indptr_last_dim}, 0));
+        indptr_indices =
+            context.mark_node(std::make_shared<v3::Broadcast>(indptr_i64, target_shape, BroadcastType::BIDIRECTIONAL));
+    }
+
+    // Create zero-padded source for prefix sum computation
+    // We need a leading zero slice along the axis so boundaries[0] = 0
+    auto element_at_zero = context.mark_node(std::make_shared<v8::Gather>(src, zero_1d, axis_const));
+    auto zero_like = context.mark_node(std::make_shared<v1::ConvertLike>(zero_scalar, src));
+    auto zero_padded = context.mark_node(std::make_shared<v1::Multiply>(element_at_zero, zero_like));
```
[the remark this answers is not in the document]
Comment (Nishant-ZFYII) [reply src/frontends/pytorch/src/op/segment_mean_csr.cpp]: I’ve updated the zero-padding path to avoid `element_at_zero * 0`, which can produce `NaN` for `Inf` inputs.  
The padding slice is now created via shape-based broadcast of a typed zero (`Broadcast(ConvertLike(0, src), ShapeOf(element_at_zero))`), so it is value-independent and numerically stable.

File: src/frontends/pytorch/src/op/segment_mean_csr.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    // Validate indptr shape
+    auto indptr_pshape = indptr.get_partial_shape();
+    PYTORCH_OP_CONVERSION_CHECK(indptr_pshape.rank().is_static(), "segment_mean_csr: indptr rank must be static");
+    auto indptr_rank = indptr_pshape.rank().get_length();
+    PYTORCH_OP_CONVERSION_CHECK(indptr_rank >= 1, "segment_mean_csr: indptr rank must be >= 1");
+
+    // Validate src shape (rank can be dynamic at conversion time)
+    auto src_pshape = src.get_partial_shape();
+
```
[the remark this answers is not in the document]
Comment (Nishant-ZFYII) [reply src/frontends/pytorch/src/op/segment_mean_csr.cpp:67]: I aligned the in-code limitation notes with the actual implementation: we only require `indptr` rank to be static, while `src` rank is allowed to be dynamic (with runtime-safe handling already in place).

File: tests/layer_tests/pytorch_tests/test_segment_mean_csr.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    TORCH_SCATTER_AVAILABLE = True
+except ImportError:
+    TORCH_SCATTER_AVAILABLE = False
+
+
+@pytest.mark.skipif(not TORCH_SCATTER_AVAILABLE, reason="torch_scatter is not installed")
+@pytest.mark.skipif(
+    PytorchLayerTest.use_torch_export() or PytorchLayerTest.use_torch_compile_backend(),
+    reason="torch_scatter::segment_mean_csr is only supported for TorchScript",
+)
```
[the remark this answers is not in the document]
Comment (Nishant-ZFYII) [reply tests/layer_tests/pytorch_tests/test_segment_mean_csr.py:20]: This is expected -- ours is the first `torch_scatter` op being added to the PyTorch Frontend (parent issue #28584), so `torch_scatter` is not yet part of the CI test requirements. The patterns we follow are consistent with the existing test suite: 
- **`try/except ImportError` + `skipif` guard** -- same approach used for optional version checks in other tests, e.g. [`test_batch_norm.py`](tests/layer_tests/pytorch_tests/test_batch_norm.py) and [`test_masked_select.py`](tests/layer_tests/pytorch_tests/test_masked_select.py) use `skipif` to handle unsupported configurations gracefully.
- **`use_torch_export()` / `use_torch_compile_backend()` skip** -- used across 13+ test files (e.g. `test_cumsum.py`, `test_mean.py`, `test_squeeze.py`, `test_eye.py`) to guard TorchScript-only paths. Adding `torch_scatter` to the CI test requirements is a broader infrastructure change that should be addressed at the parent issue level (#28584) as more `torch_scatter` ops are added, rather than in this individual PR. Tests were validated locally with `torch_scatter` installed: 8/8 CPU tests pass (FP32 + FP16).

**Comment by Nishant-ZFYII** [comment]
Thanks for the detailed review — quick status update from my side:

- **Issue 1 (HIGH) — Fixed**  
  I replaced `element_at_zero * 0` with a value-independent zero padding path using a typed zero + broadcast (`ConvertLike(0, src)` + `ShapeOf(element_at_zero)` + `Broadcast`). This avoids the `Inf * 0 -> NaN` edge case and keeps prefix boundaries numerically stable.

- **Issue 2 (MEDIUM) — Not fixed in this PR**  
  I agree with the CI coverage concern. I kept this unchanged in this PR because enabling `torch_scatter` in PT layer-test CI is a broader infra/dependency change (parent tracker: `#28584`) and should be handled there as more `torch_scatter` ops are added.

- **Issue 3 (LOW) — Fixed**  
  I aligned docs/comments with implementation behavior: we require static rank for `indptr`; `src` rank can remain dynamic at conversion time (as handled by current logic).

- **Test signature mismatch — Fixed**  
  I updated the PT test model return/signature usage to match current `PytorchLayerTest` API (`create_model()` now returns `(model, kind)`), preventing argument shifting where op kind could be interpreted as device name.

Local validation was re-run on CPU and looks good from my side.  
Happy to follow up with any additional cleanup or CI-related follow-up if you want me to take that on next.

**Comment by Nishant-ZFYII** [comment]
Good Day @mlukasze,

Can you please look into this PR. 

copilot review has been done without any comment. 

Kindly requesting you to run the workflows to see if there are any errors that I need to fix. 

Thanks and regards.

Happy to help:)

**Comment by Nishant-ZFYII** [comment]
Good Day @mlukasze ,

The CI checks have passed. 

There is one test failing and I have checked it. This failure looks unrelated to the segment_mean_csr implementation itself. pytest-html==4.2.0 exists on PyPI, but in this run pip hit a cache/index resolution issue on Windows (304 + cache read/permission problems), then failed to resolve the package and reported “No matching distribution found”. I think this is an issue in the CI setup. 

Kindly requesting you to run the jenkins build for this run.

Thanks and Regards,

Happy to help:)

**Comment by Nishant-ZFYII** [comment]
Good Day @mlukasze , @mmikolajcz , @PiotrKrzem ,

Kindly requesting you to check this PR once and let me know if there is anything I can do about the error. 

Thanks and Regards, 
Nishant.
